File: PersianCalendar/src/main/java/com/byagowi/persiancalendar/view/activity/WidgetConfigurationActivity.java
package com.byagowi.persiancalendar.view.activity;

import android.appwidget.AppWidgetManager;
import android.content.Intent;
import android.os.Bundle;

import com.byagowi.persiancalendar.R;
import com.byagowi.persiancalendar.util.UIUtils;
import com.byagowi.persiancalendar.util.UpdateUtils;
import com.byagowi.persiancalendar.util.Utils;
import com.byagowi.persiancalendar.view.preferences.FragmentWidgetNotification;

import androidx.appcompat.app.AppCompatActivity;

public class WidgetConfigurationActivity extends AppCompatActivity {
    protected void finishAndSuccess() {
        Bundle extras = getIntent().getExtras();
        if (extras != null) {
            int appwidgetId = extras.getInt(AppWidgetManager.EXTRA_APPWIDGET_ID);
            setResult(RESULT_OK, new Intent()
                    .putExtra(AppWidgetManager.EXTRA_APPWIDGET_ID, appwidgetId));
        }
        Utils.updateStoredPreference(this);
        UpdateUtils.update(this, false);
        finish();
    }

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        UIUtils.setTheme(this);
        Utils.applyAppLanguage(this);
        super.onCreate(savedInstanceState);
        setContentView(R.layout.widget_preference_layout);

        getSupportFragmentManager().beginTransaction().add(
                R.id.preference_fragment_holder,
                new FragmentWidgetNotification(), "TAG").commit();

        findViewById(R.id.add_widget_button).setOnClickListener(v -> finishAndSuccess());
    }
}

File: PersianCalendar/src/main/java/com/byagowi/persiancalendar/view/activity/WidgetConfigurationActivity.kt
package com.byagowi.persiancalendar.view.activity

import android.app.Activity
import android.appwidget.AppWidgetManager
import android.content.Intent
import android.os.Bundle
import android.view.View
import androidx.appcompat.app.AppCompatActivity
import androidx.appcompat.app.AppCompatDelegate
import androidx.preference.PreferenceFragmentCompat
import com.byagowi.persiancalendar.R
import com.byagowi.persiancalendar.util.UIUtils
import com.byagowi.persiancalendar.util.UpdateUtils
import com.byagowi.persiancalendar.util.Utils

class WidgetConfigurationActivity : AppCompatActivity() {

  // don't remove public ever
  class PreferenceFragment : PreferenceFragmentCompat() {

    override fun onCreatePreferences(savedInstanceState: Bundle?, rootKey: String?) {
      AppCompatDelegate.setCompatVectorFromResourcesEnabled(true)
      addPreferencesFromResource(R.xml.widget_preferences)
    }

    companion object {
      internal fun newInstance(): PreferenceFragment = PreferenceFragment()
    }
  }

  private fun finishAndSuccess() {
    val extras = intent.extras
    if (extras != null) {
      val appwidgetId = extras.getInt(AppWidgetManager.EXTRA_APPWIDGET_ID)
      setResult(Activity.RESULT_OK, Intent()
          .putExtra(AppWidgetManager.EXTRA_APPWIDGET_ID, appwidgetId))
    }
    Utils.updateStoredPreference(this)
    UpdateUtils.update(this, false)
    finish()
  }

  override fun onCreate(savedInstanceState: Bundle?) {
    UIUtils.setTheme(this)
    Utils.changeAppLanguage(this)
    super.onCreate(savedInstanceState)
    setContentView(R.layout.widget_preference_layout)

    supportFragmentManager.beginTransaction().add(
        R.id.preference_fragment_holder,
        PreferenceFragment.newInstance(), "TAG").commit()

    findViewById<View>(R.id.add_widget_button).setOnClickListener { finishAndSuccess() }
  }
}
